lib/stream/cache.go

```diff
@@ -0,0 +1,198 @@
+//			b.elems = b.elems[len(b.elems)-b.limit:]
+//		}
+//	}()
+//
+//	start := b.lowest
+//	end := b.endHeight()
+//	if umath.SubtractOrZero(end, height) > uint64(b.limit) {
+//		// Don't add atts that are too far behind, just ignore.
+//		return
+//	}
+//
+//	// If this is the first att, just add it.
+//	if len(b.elems) == 0 {
+//		b.elems = append(b.elems, &elem)
+//		b.lowest = height
+//
+//		return
+//	}
+//	if height < start {
+//		// Add to the front.
+//		growLen := int(start) - int(height)
+//		grow := make([]*E, growLen)
+//		grow[0] = &elem
+//		b.elems = append(grow, b.elems...)
+//		b.lowest = height
+//
+//		return
+//	} else if height > end {
+//		// Add to the back.
+//		growLen := int(height) - int(end)
+//		grow := make([]*E, growLen)
+//		grow[len(grow)-1] = &elem
+//		b.elems = append(b.elems, grow...)
+//
+//		return
+//	}
+//
+//	// Add to middle if not already present
+//	index := int(height) - int(start)
+//	b.elems[index] = &elem
+//
+//	return
+// }
+//
+// func (b *sliceCache[E]) Get(height uint64) []E {
+//	b.mu.RLock()
+//	defer b.mu.RUnlock()
+//
+//	startIndex, ok := umath.Subtract(height, b.lowest)
+//	if !ok {
+//		return nil
+//	}
+//
+//	var resp []E
+//	for i := startIndex; i < uint64(len(b.elems)); i++ {
+//		elem := b.elems[i]
+//		if elem == nil {
+//			break
+//		}
+//
+//		resp = append(resp, *elem)
+//	}
+//
+//	return resp
+// }
+
+// NewMapCache returns a new map based cache using the provided limit.
+func NewMapCache[E any](limit int) Cache[E] {
+	return &mapCache[E]{
+		limit: limit,
+		elems: make(map[uint64]E),
+	}
+}
+
+// mapCache implements Cache storing elements in a map.
+// It will only retain the highest `limit` elements.
+// It assumes elements are roughly sequential and that heights are not sparse.
+type mapCache[E any] struct {
+	limit int // Immutable
+
+	// Mutable fields
+	mu     sync.RWMutex
+	elems  map[uint64]E
+	lowest uint64 // Invalid while elems is empty
+}
+
+// Get returns all strictly sequential elements in the cache from the provided height (inclusive).
+func (c *mapCache[E]) Get(from uint64) ([]E, error) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	var resp []E
+	for height := from; ; height++ {
+		elem, ok := c.elems[height]
+		if !ok {
+			break
+		} else if height < c.lowest {
+			return nil, errors.New("height < lowest [BUG]") // Sanity check
+		}
+
+		resp = append(resp, elem)
+	}
+
+	return resp, nil
+}
+
+// Set adds the provided elements to the cache starting from the provided height (inclusive).
+// Elements are assumed to be in strictly sequential order.
+func (c *mapCache[E]) Set(from uint64, elems []E) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	for i, elem := range elems {
+		height := from + uint64(i)
+
+		if len(c.elems) >= c.limit && height < c.lowest {
+			// Don't add too low elements when cache is full, they will just be deleted below.
+			continue
+		}
+
+		if len(c.elems) == 0 {
+			c.lowest = height // Initialize lowest when first element added.
+		}
```

Comment:
Why not initializing `lowest` with max uint64 instead and remove this?

Reply:
Sure, we can do that: `c.lowest` is the lower height in the cache or `math.Maxuint64` while the cache is empty.